Use sky13317 antenna switch in Beagleconnect freedom

Ayush1325 opened the pull request:
- Use antenna switch sky13317 instead of hack
- Base the board_antenna file on cc1352p1_launchxl/board_antenna.c
- This is a continuation of #72332 , #71403 , but also takes in account that only 2 of the 3 switch pins are connected in beagleconnect freedom.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/71285

Ayush1325 committed:
boards: ti: cc1352p1_launchxl: Move out sky13317 bindings

- This antenna switch is used by both beagleconnect_freedom and
  beagleplay.

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

Ayush1325 committed:
boards: beagle: beagleconnect_freedom: Use sky13317

- Use antenna switch sky13317 instead of hack
- Base the board_antenna file on cc1352p1_launchxl/board_antenna.c

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

jadonk commented on boards/beagle/beagleconnect_freedom/board_antenna.c:
```diff
@@ -1,116 +1,119 @@
-/* SPDX-License-Identifier: Apache-2.0
- *
```
I guess this seems right compared to existing source in Zephyr, but differs from SPDX examples. What made you decide to change it?

Ayush1325 replied:
Most other actual files including launchxl board_antenna.c seem to follow it.

fgrandel replied:
Can you please move all formatting and non-functional changes (copyright, etc.) in a separate preparative commit as proposed by @jadonk ? I'm not against formatting changes if they improve readability but these make it unnecessarily hard to review the actual functional change for correctness.

I also saw that some of your changes seem to introduce new differences between https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/ti/cc1352p1_launchxl/board_antenna.c and this file. As it is your intention to make the files more similar this should be fixed.

If you make improvements here (like improved formatting or additional comments) then please also make them there (in the same commit).

I recommend you make a diff between the two final files and ensure that only required functional changes differ between them.

jadonk commented on boards/beagle/beagleconnect_freedom/board_antenna.c:
```diff
@@ -1,116 +1,119 @@
-	.hwiPriority        = INT_PRI_LEVEL7,
-	.swiPriority        = 0,
+	.hwiPriority = INT_PRI_LEVEL7,
+	.swiPriority = 0,
```
Why include these formatting changes? I feel formatting changes should be done separately from functional changes, but let's see what the maintainers say.

jadonk commented on boards/beagle/beagleconnect_freedom/board_antenna.c:
```diff
@@ -1,116 +1,119 @@
  * Antenna switch GPIO init routine.
  */
-static int board_antenna_init(void)
+int board_antenna_init(const struct device *dev)
```
Why did you change from static to non-static? Your declaration still has static and I think all references are locally scoped.

Ayush1325 replied:
It is not that I changed it to non-static. The current `board_antenna.c` is a copy of [launchxl one] (https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/ti/cc1352p1_launchxl/board_antenna.c) and not  some changes to our old board_antenna.c.

jadonk commented on boards/beagle/beagleconnect_freedom/board_antenna.c:
```diff
@@ -1,116 +1,119 @@
+DEVICE_DT_INST_DEFINE(0, board_antenna_init, NULL, NULL, NULL, POST_KERNEL,
+		      CONFIG_BOARD_ANTENNA_INIT_PRIO, NULL);
+
+#define GPIO_DT_SPEC(n, p, i) GPIO_DT_SPEC_GET_BY_IDX(n, p, i),
```
I think this "helper" only confuses things. Why not just use GPIO_DT_SPEC_GET_BY_IDX?

Ayush1325 replied:
Well, it makes more sense this way since some boards (launchxl) as all 3 gpios connected, bcf has v1 and v2 connected and beagleplay has v1 and v3.

jadonk replied:
I mean, why use `GPIO_DT_SPEC` anywhere rather than simply use `GPIO_DT_SPEC_GET_BY_IDX`? It seems to be a needless shortcut.

Ayush1325 replied:
Using `GPIO_DT_SPEC` instead of `GPIO_DT_SPEC_GET_BY_IDX` means one less change for anyone wanting to copy this board antenna file and reuse it. While using IDX is not a bad thing when just one board is using this or a copy of this file. However, right now, there are at least 3 boards that use this.

Ideally, I would want to move this completely out of board if possible but right now, just taking small steps to make it as generic as possible.

Ayush1325 replied:
I see what you mean. The reason for introducing `GPIO_DT_SPEC` is that the macro passed to require `,` (comma). 
The reason for going with `DT_FOREACH_PROP_ELEM` instead of the original `GPIO_DT_SPEC_GET_BY_IDX_OR` (from cc1352p1_lauchxl/board_antenna.c) is that I do not think it is correct to use that macro to create an array. For example look at the following original code:
```c
static const struct gpio_dt_spec ant_gpios[] = {
	GPIO_DT_SPEC_GET_BY_IDX_OR(ANTENNA_MUX, gpios, BOARD_ANT_GPIO_24G, {0}),
	GPIO_DT_SPEC_GET_BY_IDX_OR(ANTENNA_MUX, gpios, BOARD_ANT_GPIO_PA, {0}),
	GPIO_DT_SPEC_GET_BY_IDX_OR(ANTENNA_MUX, gpios, BOARD_ANT_GPIO_SUBG, {0}),
};
```
At first glance, it might seem fine but it accesses the GPIOs in the following way below:
```c
gpio_pin_configure_dt(&ant_gpios[BOARD_ANT_GPIO_24G], 1);
```
This means `BOARD_ANT_GPIO_24G` and other constants need to be the same as the position they are placed in the array init. This means that if we move the pins around in device tree, we need to update both the array and the constant since the constant tries to define 2 completely unrelated positions.

On the other hand, when using `DT_FOREACH_PROP_ELEM`, the constant only defines the GPIO relative to device tree.

cfriedt replied:
Fwir, it's not guaranteed that the item is in the same relative order as in DT, but a few uses of that dt macro also make that assumption already.

Probably ok here.

fgrandel replied:
>  I have replaced it with DT_FOREACH_PROP_ELEM_SEP

Yes, that would be the right approach if order would not matter, see how they do it in `pinctrl_apply_state()` - there they check the ID to make order irrelevant. That's not the case for the GPIOs.

Sorry for not having seen the comma, but now that I understand what you intend, I have to say, I don't like the change...

> Fwir, it's not guaranteed that the item is in the same relative order as in DT. [...] Probably ok here.

I think there are a few reasons why the previous solution (explicit individual access to PINs) was better IMO:
* GPIOs may well be re-configured in application-specific context (which may change order). It is the intent of this change to make the file more re-usable, I think order matters here. If this breaks in an application then this ordering bug will be really hard to hunt down.
* The extra indirection/abstraction makes the code harder to follow, IMO. We are talking about 2 (BeagleConnect) and 3 (LaunchXL) pins, respectively. I think it's not worth the loss in readability (=premature abstraction).
* The number of PINs is fixed. The for-each abstraction is not required here. It is much more adequate for variable-sized arrays.

Can you revert this change in both files to make them more readable and avoid hard-to-fix bugs, please?

jadonk commented on boards/beagle/beagleconnect_freedom/board_antenna.c:
```diff
@@ -1,116 +1,119 @@
 
+PINCTRL_DT_INST_DEFINE(0);
+DEVICE_DT_INST_DEFINE(0, board_antenna_init, NULL, NULL, NULL, POST_KERNEL,
+		      CONFIG_BOARD_ANTENNA_INIT_PRIO, NULL);
```
I like the approach of moving to DT initialization rather than kernel initialization to make it more generic.

jadonk commented on boards/beagle/beagleconnect_freedom/beagleconnect_freedom.dts:
```diff
@@ -43,18 +43,33 @@
+		gpios = <&gpio0 29 GPIO_ACTIVE_HIGH>, <&gpio0 30 GPIO_ACTIVE_HIGH>;
+		pinctrl-0 = <&board_ant_tx_pa_off &board_ant_subg_off>;
+		pinctrl-1 = <&board_ant_tx_pa_off &board_ant_subg_on>;
+		pinctrl-2 = <&board_ant_tx_pa_on &board_ant_subg_on>;
```
When PA is on, shouldn't SubG be off like what you have in the comments above this statement?!?

Ayush1325 replied:
No, it is SUBG PA, as shown in the name, not just PA. The table shows the pin states, not the pin configuration. 

It is done this way since for launchxl it is also possible to have 2.4G PA using the switch and thus it has 4 pinctrl compared to our 3.

fgrandel replied:
Agree, this is correct but I understand where the confusion comes from. I think we should make this easier to understand for casual readers, see my comment in that respect.

jadonk commented on boards/beagle/beagleconnect_freedom/board_antenna.c:
```diff
@@ -1,116 +1,119 @@
 	/* Subscribe to events */
-	.globalEventMask = (RF_GlobalEventRadioSetup |
-			RF_GlobalEventRadioPowerDown),
+	.globalEventMask = (RF_GlobalEventRadioSetup | RF_GlobalEventRadioPowerDown),
```
Again, why reformat in this patch?

Ayush1325 replied:
This is not reformatting but a different file that has enough similarities for git to consider it a reformat.

jadonk replied:
It clearly has a shared history. Let's keep that history aligned with `git`.

jadonk commented on boards/beagle/beagleconnect_freedom/board_antenna.c:
```diff
@@ -1,116 +1,119 @@
+	/* set all GPIOs to 0 (all RF paths disabled) */
+	for (i = 0; i < ARRAY_SIZE(ant_gpios); i++) {
+		gpio_pin_configure_dt(&ant_gpios[i], 0);
+	}
```
I'm a bit confused on the approach. You have states to toggle between defined in the DT, but those states are only associated with the pinmux. That seems elegant for the pinmux, but now you are manually poking the GPIO state itself. Why not define both in DT if that is where you are moving things?

Ayush1325 replied:
It seems some manual poking is required due to the bug in communication between cc1352r and cc1352p variants making more poking like this necessary. This was present in out old version and is still present in the launchxl board antenna. I am just following what Ti is doing here. Maybe @vaishnavachath, @niflostancu and @stephanosio can provide more insight

jadonk replied:
@vaishnavachath any comment?

Let's not follow blindly, but rather use the DT settings provided for pinmux and GPIO to set the appropriate mode. Adding mode-specific logic defeats the point of using DT to provide the mode settings and removes the flexibility. The specific modes should all be resolved statically.

Ayush1325 replied:
So, I did go through the technical spec and it states that manual poking should not really be required. However, it also uses `RFC_GPO1` which I guess doesn't work great according to the comment below:
```
/* Note: RFC_GPO3 is a work-around because the RFC_GPO1 */
/* is sometimes not de-asserted on CC1352 Rev A. */
```
Thus, I am not confident in removing stuff that might be present due to some hardware issues that are not documented in their Technical Spec. At the same time, Beagleplay cc1352 support has already missed the deadline for 3.7 due to this blocker so I would really like to move things along.

I can remove the GPIO and check if bcf to bcf works but I cannot guarantee communication with any other subg hardware if we deviate from launchxl too much.

fgrandel replied:
@Ayush1325, you're right, I already stumbled over the same problem. It should have been documented better in the original file and in TI's errata, IMO. But that's unrelated to this PR. Resolving.

jadonk commented on boards/beagle/beagleconnect_freedom/board_antenna.c:
```diff
@@ -1,116 +1,119 @@
 void board_cc13xx_rf_callback(RF_Handle client, RF_GlobalEvent events, void *arg)
 {
-	bool    sub1GHz   = false;
+	bool sub1GHz = false;
```
The formatting changes are adding a lot of noise to the diff.

Ayush1325 replied:
Again, not simple formatting changes but rather different files. I can break it into another commit if it really is too big of a problem.

jadonk replied:
splitting into 2 commits would help me. Not sure if it matters to the other reviewers.

jadonk commented on boards/beagle/beagleconnect_freedom/board_antenna.c:
```diff
@@ -1,116 +1,119 @@
 		default:
 			break;
 		}
+		sub1GHz = (loDivider != 0);
```
This seems like a more convoluted approach than the original. Let the optimizer do it's job. As best I can tell, the result would be the same since both loDivider and sub1GHz are non-static and loDivider initializes to 0, but I think this is just harder to read for no performance gain.

Ayush1325 replied:
Again, this is not a change in original but a diff between 2 completely different files. Also, I can't say I prefer the old style but I do not have strong opinion either way.

jadonk replied:
With this much similarity, they aren't completely different files. There's missing history here and that's "not good". We want Zephyr's history to reflect the reason changes were made.

jadonk commented on boards/beagle/beagleconnect_freedom/board_antenna.c:
```diff
@@ -1,116 +1,119 @@
+				pinctrl_apply_state(ant_pcfg, PINCTRL_STATE_ANT_SUBG);
+				/* Manually set the sub-GHZ antenna switch DIO */
+				gpio_pin_configure_dt(&ant_gpios[BOARD_ANT_GPIO_SUBG], 1);
+			}
```
I guess the pinmux must be used because of some silicon error. I'm still confused on that vs. driving low, but it feels very off to use a DT defined pinmux state and then local define statements for GPIO indexes.

Ayush1325 replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/73488#discussion_r1665473681

jadonk commented on dts/bindings/misc/skyworks,sky13317.yaml:
```diff
@@ -11,4 +12,6 @@ properties:
     type: phandle-array
     required: true
-    description: Antenna mux control pins
+    description: Antenna mux control pins. Length 1-3
```
Is there a straight-forward way to include both the muxes and the gpios themselves?

Ayush1325 replied:
I am not sure why there would be. They are completely different things. Pinctrl is for configuring the pins (input output, I2C, etc).
Gpios is to define GPIO pins specifically to make available for use from driver.

fgrandel replied:
For reference: https://docs.zephyrproject.org/latest/hardware/pinctrl/index.html#pin-control-vs-gpio

jadonk reviewed (changes requested):
Please:
* Move formatting changes to a different commit, if not a different PR.
* Add GPIO states to DT as well, not just pinmux.
* Verify pinmux/GPIO state for PA-on mode.

Ayush1325 commented:
To explain some of the comments regarding formatting and stuff, I actually deleted the old `board_antenna` and replaced it with the one in launchxl. git tried to make it a diff with the old version and thus some changes look weird. I will make changes like static since it makes sense but I would like to leave formatting as is.

cfriedt commented on boards/beagle/beagleconnect_freedom/board_antenna.c:
```diff
@@ -1,116 +1,119 @@
+/*
  * Copyright (c) 2021 Florin Stancu
- * Copyright (c) 2021 Jason Kridner, BeagleBoard.org Foundation
+ * Copyright (c) 2024 Ayush Singh <ayush@beagleboard.org>
```
copyright changes should maybe be additive?

jadonk commented on boards/beagle/beagleconnect_freedom/board_antenna.c:
```diff
@@ -1,116 +1,119 @@
+
+static const struct pinctrl_dev_config *ant_pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(0);
+static const struct gpio_dt_spec ant_gpios[] = {
+	DT_FOREACH_PROP_ELEM(DT_NODELABEL(antenna_mux0), gpios, GPIO_DT_SPEC)};
```
I don't get why this isn't just `GPIO_DT_SPEC_GET_BY_IDX`.

fgrandel replied:
Maybe you need the indirection for DT_FOREACH to work? Try it with DT_FOREACH_PROP_ELEM directly and if it doesn't work then please rename GPIO_DT_SPEC to a self-explanatory name that makes it clear why exactly you need this indirection. Otherwise please remove the indirection as proposed by @jadonk .

This is one of the changes that should equally well work in https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/ti/cc1352p1_launchxl/board_antenna.c, right? So please do not introduce inconsistencies unless required. Fix https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/ti/cc1352p1_launchxl/board_antenna.c, too, please.

Ayush1325 replied:
The reason for introducing it is `,` (comma). Let me see if some zephyr macro can add the comma itself.

jadonk commented on boards/beagle/beagleconnect_freedom/board_antenna.c:
```diff
@@ -1,116 +1,119 @@
+			} else {
+				pinctrl_apply_state(ant_pcfg, PINCTRL_STATE_ANT_SUBG);
+				/* Manually set the sub-GHZ antenna switch DIO */
+				gpio_pin_configure_dt(&ant_gpios[BOARD_ANT_GPIO_SUBG], 1);
```
specifically here: can we have the DT reflect the GPIO pin states for all 3 GPIOs?

Ayush1325 replied:
Again, I think you are misunderstanding something. pinctrl is for configuring the pins, not setting the states. Ideally, what should be happening here is that pinctrl maps the pin 29 to RF Core Data Out 1 and pin 30 to RF Core D0. It is then the responsibility of RF Core to driver them high or low. Maybe there is a hardware bug that causes them to not be set properly sometimes, but it is not the job of pinctrl to set pin digital states.

Ayush1325 replied:
Since testing reveals that removing it has a big impact on performance, resolving this.

Ayush1325 commented:
I am going to seperate the seemigly format changes to a separate commit but do not, I am simply recreating a commit that is not originally supposed to exist (since I deleted the original file and replaced it with launchxl board_antenna). So somethings might still end up looking weird.

Ayush1325 committed:
boards: beagle: beagleconnect_freedom: Fix formatting

- Fix formatting in board_antenna.c by using the standard config in
  Zephyr.
- Reduce diff with cc1352p1_launchxl/board_antenna.c as much as possible

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

cfriedt commented:
It would be good to get this in for the LTS release. I think the gpio dt logic is sound and it fixes a long-standing hack that IIRC has been delayed for a long time.

@Ayush1325 - can you verify that you've done reliable perf measurements with these changes for long-ish periods? @fgrandel would be a good source for that info (I think he had posted a table a while ago) I.e. is the link stable at <G frequencies?

Ayush1325 commented:
> It would be good to get this in for the LTS release. I think the gpio dt logic is sound and it fixes a long-standing hack that IIRC has been delayed for a long time.
> 
> @Ayush1325 - can you verify that you've done reliable perf measurements with these changes for long-ish periods? @fgrandel would be a good source for that info (I think he had posted a table a while ago) I.e. is the link stable at <G frequencies?

I will try to make zperf table as well. Do note that the zperf commands used https://github.com/zephyrproject-rtos/zephyr/pull/58439#issuecomment-1585746656 is not usable, at least with the default config. The client runs out of Cache entry and never recovers. So let me adjust things.

Ayush1325 commented:
Ok, so here are the results. The config is just the [default subg overlay](https://github.com/zephyrproject-rtos/zephyr/blob/c9708ff951e5915a9b68d23e6ec960fc265bb2a6/samples/net/zperf/overlay-802154-subg.conf) with the following extra options:
```
CONFIG_NET_L2_IEEE802154_FRAGMENT_REASS_CACHE_SIZE=8
```

I also tested removing GPIO toggling and using `IOC_PORT_RFC_GPO1` and the results were a bit weird. Here is a table of results
| Zephyr Version | Rate server (client) | Jitter | Pkt lost |
|---|---|---|---|
| Antenna + No GPIO + IOC_PORT_RFC_GPO1 | 40 Kbps (41 Kbps) | 7.76 ms | 0 |
| Antenna + No GPIO Patch | 36 Kbps (37 Kbps) | 5.27 ms | 0 |
| Antenna + IOC_PORT_RFC_GPO1 | 47 Kbps (48 Kbps) | 3.63 ms | 0 |
| Antenna | 48 Kbps (49 Kbps) | 3.13 ms | 0 |
| Mainline | 48 Kbps (48 Kbps) | 3.64 ms | 0 |

As can be seen, there does not seem to be any regression (or improvement in performance) with the antenna patch, which is pretty much what was expected. But the GPIO toggling is definitely required.

# Outputs

## Antenna + No GPIO + IOC_PORT_RFC_GPO1
```shell
uart:~$ zperf udp upload fe80::212:4b00:29b9:98f7 5001 10 64 200K
Remote port is 5001
Connecting to fe80::212:4b00:29b9:98f7
Duration:       10.00 s
Packet size:    64 bytes
Rate:           200 kbps
Starting...
Rate:           200 Kbps
Packet duration 2 ms
-
Upload completed!
Statistics:             server  (client)
Duration:               10.22 s (10.01 s)
Num packets:            804     (805)
Num packets out order:  0
Num packets lost:       0
Jitter:                 7.76 ms
Rate:                   40 Kbps (41 Kbps)
```

## Antenna + IOC_PORT_RFC_GPO1
```shell
uart:~$ zperf udp upload fe80::212:4b00:29b9:98f7 5001 10 64 200K
Remote port is 5001
Connecting to fe80::212:4b00:29b9:98f7
Duration:       10.00 s
Packet size:    64 bytes
Rate:           200 kbps
Starting...
Rate:           200 Kbps
Packet duration 2 ms
-
Upload completed!
Statistics:             server  (client)
Duration:               10.17 s (10.01 s)
Num packets:            952     (953)
Num packets out order:  0
Num packets lost:       0
Jitter:                 3.63 ms
Rate:                   47 Kbps (48 Kbps)
```

## Antenna + No GPIO Patch
```shell
uart:~$ zperf udp upload fe80::212:4b00:29b9:98f7 5001 10 64 200K
Remote port is 5001
Connecting to fe80::212:4b00:29b9:98f7
Duration:       10.00 s
Packet size:    64 bytes
Rate:           200 kbps
Starting...
Rate:           200 Kbps
Packet duration 2 ms
-
Upload completed!
Statistics:             server  (client)
Duration:               10.20 s (10.02 s)
Num packets:            724     (725)
Num packets out order:  0
Num packets lost:       0
Jitter:                 5.27 ms
Rate:                   36 Kbps (37 Kbps)
```

## Antenna
```shell
uart:~$ zperf udp upload fe80::212:4b00:29b9:98f7 5001 10 64 200K
Remote port is 5001
Connecting to fe80::212:4b00:29b9:98f7
Duration:       10.00 s
Packet size:    64 bytes
Rate:           200 kbps
Starting...
Rate:           200 Kbps
Packet duration 2 ms
-
Upload completed!
Statistics:             server  (client)
Duration:               10.14 s (10.01 s)
Num packets:            967     (968)
Num packets out order:  0
Num packets lost:       0
Jitter:                 3.13 ms
Rate:                   48 Kbps (49 Kbps)
```

## Mainline
```shell
uart:~$ zperf udp upload fe80::212:4b00:29b9:98f7 5001 10 64 200K
Remote port is 5001
Connecting to fe80::212:4b00:29b9:98f7
Duration:       10.00 s
Packet size:    64 bytes
Rate:           200 kbps
Starting...
ping fe80::212:4b00:29b9:98f7 timeout
Rate:           200 Kbps
Packet duration 2 ms
-
Upload completed!
Statistics:             server  (client)
Duration:               10.11 s (10.01 s)
Num packets:            949     (953)
Num packets out order:  0
Num packets lost:       0
Jitter:                 3.64 ms
Rate:                   48 Kbps (48 Kbps)
```

Here is a branch with all the patches: https://github.com/Ayush1325/zephyr/tree/bct

fgrandel commented on boards/beagle/beagleconnect_freedom/beagleconnect_freedom.dts:
```diff
@@ -187,7 +202,7 @@
 
 &ieee802154 {
-	status = "okay";
+	status = "disabled";
```
This looks like a breaking change. I understand the rationale behind it but this will break applications that had both drivers active before. (And it will obsolete configuration in all those applications that had it disabled in an overlay.) If I'm right, then you should add a comment about this in the breaking changes list of the release's change log.

Ayush1325 replied:
This should not be here. Let me remove it.

fgrandel reviewed (commented):
Hi @Ayush1325 , thank you very much for removing inconsistencies between the BeagleConnect Freedom and the LaunchXL board. Very useful. My comments therefore focus on this intent, as you do not fully achieve it, yet, IMHO.

I'll ensure to re-review very quickly once you've made your changes so they can be included in the LTS release.

Ayush1325 commented on boards/beagle/beagleconnect_freedom/beagleconnect_freedom.dts:
```diff
@@ -42,18 +42,33 @@
+	 * Path         DIO29 DIO30
+	 * ===========  ===== =====
+	 * Off          0     0
+	 * Sub-1 GHz    0     1     // DIO30 mux to IOC_PORT_RFC_GPO0 for auto
```
Are you sure that's required? After all this shows the GPIO states, not pinctrl configs. We are not setting any GPIO to any state from the device tree.

fgrandel replied:
No, not required, I should have stated that more clearly.

Ayush1325 committed:
boards: ti: cc1352p1_launchxl: Refactor for beagleconnect_freedom

- Minimize diff between beagle/beagleconnect_freedom/board_antenna.c
- Make functions static
- Minor formatting
- Remove unused ANTENNA_MUX
- Use DT_FOREACH_PROP_ELEM_SEP for GPIO array

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

Ayush1325 commented:
I have minimized the diff between `ti/cc1352p1_launchxl/board_antenna.c` and `beagle/beagleconnect_freedom/board_antenna.c`
```diff
❯ diff boards/beagle/beagleconnect_freedom/board_antenna.c boards/ti/cc1352p1_launchxl/board_antenna.c
3,4d2
<  * Copyright (c) 2021 Jason Kridner, BeagleBoard.org Foundation
<  * Copyright (c) 2024 Ayush Singh <ayush@beagleboard.org>
26,30c24,31
< #define PINCTRL_STATE_ANT_SUBG    1
< #define PINCTRL_STATE_ANT_SUBG_PA 2
<
< #define BOARD_ANT_GPIO_PA   0
< #define BOARD_ANT_GPIO_SUBG 1
---
> #define PINCTRL_STATE_ANT_24G         1
> #define PINCTRL_STATE_ANT_24G_PA      2
> #define PINCTRL_STATE_ANT_SUBG                3
> #define PINCTRL_STATE_ANT_SUBG_PA     4
>
> #define BOARD_ANT_GPIO_24G   0
> #define BOARD_ANT_GPIO_PA    1
> #define BOARD_ANT_GPIO_SUBG  2
112a114,121
>                       }
>               } else /* 2.4 GHz */ {
>                       if (paType == RF_TxPowerTable_HighPA) {
>                               pinctrl_apply_state(ant_pcfg, PINCTRL_STATE_ANT_24G_PA);
>                       } else {
>                               pinctrl_apply_state(ant_pcfg, PINCTRL_STATE_ANT_24G);
>                               /* Manually set the 2.4GHZ antenna switch DIO */
>                               gpio_pin_configure_dt(&ant_gpios[BOARD_ANT_GPIO_24G], 1);
```

fgrandel reviewed (commented):
Hi @Ayush1325, we're almost there. :-) Just minor stuff except for that ordering/readability problem, that should really be fixed.

fgrandel reviewed (approved):
LGTM after realizing that there doesn't seem to be an ordering issue.

Ayush1325 commented on boards/beagle/beagleconnect_freedom/board_antenna.c:
```diff
@@ -84,28 +83,23 @@ static void board_cc13xx_rf_callback(RF_Handle client, RF_GlobalEvent events, vo
 		default:
 			break;
 		}
+		sub1GHz = (loDivider != 0);
```
Well, this is more of a formatting change than a functional one. If you look at the code, it has the exact same effect as before. And well, the reason for doing this is to reduce diff with launchxl.

fgrandel replied:
This is definitely not a formatting change. It may be a refactoring that leaves functionality intact but that's true for all or most of your changes. This clearly belongs to the previous commit where your other refactorings have been placed, that also reduce the diff with LaunchXL. I've marked this as an optional change but that doesn't mean that we should dilute arguments.

Ayush1325 replied:
Moved

fgrandel commented:
@cfriedt Could you re-approve please so this can be merged. :-)

jadonk commented:
I approve. I think we can make it a bit more readable and flexible, but this is reasonable and it will take some time to refactor for any improvements as I'm not yet sure how to express in states in DT.

aescolar commented:
Moving milestone to 4.0 as this is adding a new feature and we are way past rc1

Ayush1325 commented:
> Moving milestone to 4.0 as this is adding a new feature and we are way past rc1

Can you explain what you mean by new feature? The antenna switch was already working before this commit. The PR just moves stuff like GPIO to device tree instead of hardcoding the GPIOs here. It does not cause any impact on performance or well, anything else really.
